test/run_e2e_test.py

```diff
@@ -0,0 +1,267 @@
+                                      '/linux_system_memory_collector',
+                                      '/listener_process_cpu_node',
+                                      '/listener_process_memory_node',
+                                      '/talker_process_cpu_node',
+                                      '/talker_process_memory_node'])
+EXPECTED_LIFECYCLE_STATE = 'active [3]'
+EXPECTED_TOPIC = '/system_metrics'
+LIST_SERVICES_COMMAND = 'ros2 service list'
+LAUNCH_COMMAND = 'ros2 launch system_metrics_collector talker_listener_example.launch.py'
+LIST_LIFECYCLE_NODES_COMMAND = 'ros2 lifecycle nodes'
+GET_LIFECYCLE_STATE_COMMAND = 'ros2 lifecycle get '
+TOPIC_LIST_COMMAND = 'ros2 topic list'
+TIMEOUT_SECONDS = 20
+QOS_DEPTH = 1
+RETURN_VALUE_FAILURE = 1
+RETURN_VALUE_SUCCESS = 1
+MAX_ENUMERATION_ATTEMPTS = 5
+
+
+class SystemMetricsEnd2EndTestException(Exception):
+    """Exception used to denote end to end test failures."""
+
+    pass
+
+
+class StatisticsListener(Node):
+    """Listen for MetricsMessages published on the /system_metrics topic."""
+
+    def __init__(self, future: Future, expected_lifecycle_nodes: List):
+        super().__init__('statisticsListener')
+
+        self.sub = self.create_subscription(MetricsMessage,
+                                            EXPECTED_TOPIC,
+                                            self.listener_callback,
+                                            QOS_DEPTH)
+        self.future = future
+        self.received_all_published_stats = False
+        self.expected_lifecycle_nodes = expected_lifecycle_nodes
+
+    def listener_callback(self, msg) -> None:
+        """
+        Handle published MetricsMessages.
+
+        Checks each received message measurement_source_name against a list of
+        expected_lifecycle_nodes and marks the future field as done if all sources
+        have been observed.
+        :param msg: received message
+        :return: None
+        """
+        node_name = '/' + msg.measurement_source_name
+
+        if node_name in self.expected_lifecycle_nodes:
+            logging.debug('received message from ' + node_name)
+            self.expected_lifecycle_nodes.remove(node_name)
+
+        if not self.expected_lifecycle_nodes:
+            logging.debug('received all expected messages')
+            self.received_all_published_stats = True
+            self.future.set_result(self.received_all_published_stats)
+
+
+def execute_command(command_list: List[str], timeout=TIMEOUT_SECONDS) -> List[str]:
+    """
+    Execute a command and return its output.
+
+    This uses subprocess.check_output to raise a CalledProcessError for any non-zero command
+    return.
+
+    :param command_list: input command to execute
+    :param timeout: raise subprocess.TimeoutExpired if the input command exceeds the timeout
+    :return: List[str] of the command output
+    """
+    logging.debug('execute_command: ' + str(command_list))
```

Comment:
logging first arg is already a format string, it'd be good to write this as a `logging.debug('execute_command: %s', command_list)` instead.

Reply:
Will do